File: static/js/nfl_qbs.js
function get_rolling_epa_and_cpoe(qb_stats) {
  let d1 = [];
  let d2 = [];
  d1.push({
    label: 'EPA',
    fill: false,
    borderColor: qb_stats.colors[0],
    backgroundColor: qb_stats.colors[0],
    pointRadius: 3,
    data: qb_stats.epa
  });

  d2.push({
    label: 'CPOE',
    fill: false,
    borderColor: qb_stats.colors[0],
    backgroundColor: qb_stats.colors[0],
    pointRadius: 3,
    data: qb_stats.cpoe
  });


  var epa_threshold = Array(qb_stats.games.length).fill(qb_stats.league_avg_epa);
  epaT = {
    label: "League Average EPA per Game",
    fill: false,
    borderColor: "#ff0000",
    backgroundColor: "#ff0000",
    pointRadius: 0,
    data: epa_threshold
  };

  var cpoe_threshold = Array(qb_stats.games.length).fill(qb_stats.league_avg_cpoe);
  cpoeT = {
    label: "League Average CPOE per Game",
    fill: false,
    borderColor: "#ff0000",
    backgroundColor: "#ff0000",
    pointRadius: 0,
    data: cpoe_threshold
  };

  d1.push(epaT);
  d2.push(cpoeT);


  let rolling_epa_avg = line_chart('rolling_epa_avg',
                                qb_stats.games,
                                d1,
                                qb_stats.name + ' Rolling 5 game EPA Average',
                                'regular season only',
                                {}, {},
                                'NFLFastR')
  let rolling_cpoe_avg = line_chart('rolling_cpoe_avg',
                                qb_stats.games,
                                d2,
                                qb_stats.name + ' Rolling 5 game CPOE Average',
                                'regular season only',
                                {}, {},
                                'NFLFastR')
}

File: nfl_qbs.py
import pandas as pd
import json
import helper as help

def get_rolling_epa_and_cpoe(player, seasons):
    df = pd.DataFrame()
    for s in seasons:
        df = df.append(help.get_nfl_data(s))

    #regular season data only
    df = df.loc[df.season_type=='REG']

    qbs = df.copy()
    qbs = qbs.groupby(['passer', 'posteam', 'game_id'], as_index=False).agg({'epa':'mean',
                                                              'cpoe':'mean',
                                                              'play_id':'count'})
    #Filter to players with 10 or more dropbacks in a game
    qbs = qbs.loc[qbs.play_id>9]
    #get league average epa and cpoe per game
    qbs = qbs.agg({'epa':'mean', 'cpoe':'mean'})


    df = df.loc[df.passer==player]
    df = df.groupby(['posteam', 'passer', 'game_id'], as_index=False).agg({'epa':'mean',
                                                              'cpoe':'mean'})

    games = []
    epa = []
    cpoe = []
    for index, row in df.iterrows():
        r = row.to_dict()
        game_data = r.get('game_id').split('_')
        games.append(game_data[0] + ' Week ' + game_data[1] + ': ' + game_data[2] + ' @ ' + game_data[3])
        epa.append(r.get('epa'))
        cpoe.append(r.get('cpoe'))
        team = r.get('posteam')

    return {
        'name': player,
        'games': games,
        'epa': help.get_5_game_rolling_average(epa),
        'cpoe': help.get_5_game_rolling_average(cpoe),
        'colors': help.get_nfl_team_colors(team),
        'league_avg_epa' : qbs.get(key='epa'),
        'league_avg_cpoe' : qbs.get(key='cpoe')
    }

def get_qb_epa_ranked(seasons):
    df = pd.DataFrame()
    for s in seasons:
        df = df.append(help.get_nfl_data(s))

    qbs = df.groupby(['passer', 'posteam'], as_index=False).agg({'epa':'mean',
                                                              'play_id':'count'})
    #Filter to players with 200 or more dropbacks in a game
    qbs = qbs.loc[qbs.play_id>199]
    qbs.sort_values('epa', ascending=False, inplace=True)

    passers = []
    epa = []
    colors = []
    for index, row in qbs.iterrows():
        r = row.to_dict()
        passers.append(r.get('passer'))
        epa.append(r.get('epa'))
        team = r.get('posteam')
        colors.append(help.get_nfl_team_colors(team)[0])

    return {
        'qbs': passers,
        'epa': epa,
        'colors': colors,
        'situation': 'Regular Season Only',
        'seasons': str(seasons[0]) if len(seasons) == 1 else str(seasons[0]) + ' - ' + str(seasons[len(seasons)-1])
    }


#if __name__ == '__main__':
#    player = 'C.Wentz'
#    seasons = list(range(2020, 2021)) #note end is not inclusive
#    r = get_rolling_epa_and_cpoe(player, seasons)
#    r = get_qb_epa_ranked(seasons)
#    print(str(r))
